compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,410 @@
+ * ```
+ *
+ * The checker first checks the base types `ArrayList` and `List`, and it sees no issue since `ArrayList <: List`. Next it checks the type
+ * parameters bound to these base types: `T1` in `ArrayList<T1>` and `T2` in `List<T2>`. For `T1`, there is only one bound type argument
+ * `Set<String>`, so it's good. For `T2`, there are two bound type arguments: `Set<String>` and `HashSet<Int>`. Now the checker recursively
+ * checks whether these two types are compatible. Again, it first checks the base type `Set` and `HashSet`, and it finds no problem since
+ * `HashSet <: Set`. Finally, checks the type arguments `String` and `Int` that are bound to `T` in `Set<T>`. They are incompatible since
+ * `String` and `Int` are unrelated classes.
+ *
+ * The above example only goes over covariant type arguments. For contravariant types, the checker simply checks whether the range formed by
+ * covariant and contravariant bounds is empty. For example, a range like `[Collection, List]` is empty and hence invalid because `List` is
+ * not a super class/interface of `Collection`
+ */
+object ConeTypeCompatibilityChecker {
+    private val ConeInferenceContext.mimicOldFrontendBehavior: Boolean
+        get() = !session.languageVersionSettings.supportsFeature(LanguageFeature.ExtendedTypeCompatibilityChecker)
+
+    fun Collection<ConeKotlinType>.areCompatible(ctx: ConeInferenceContext): Boolean {
+        // If all types are nullable, then `null` makes the given types compatible.
+        if (all { with(ctx) { it.isNullableType() } }) return true
+        // Next can simply focus on the type hierarchy and don't need to worry about nullability.
+        return ctx.areCompatible(flatMap { it.collectUpperBounds() }.toSet(), emptySet())
+    }
+
+    private fun ConeInferenceContext.areCompatible(upperBounds: Set<ConcreteType>, lowerBounds: Set<ConcreteType>): Boolean {
+        val upperBoundConcreteClassToTypes: Map<ConcreteClass, List<ConcreteType>> =
+            upperBounds.groupBy { it.toConcreteClass(this) }.filterKeys { it != null }.cast()
+
+        // Map all types to the upper bound class in the class hierarchy and check the leaf classes. We can know the types are incompatible
+        // if
+        //   * two or more are not interfaces, or
+        //   * one class is final and there are more than one distinct interfaces
+
+        // Put the upper bound classes into the class hierarchy and count how many subclasses are there for each encountered class
+        val subClassCounts: MutableMap<ConcreteClass, AtomicInteger> = mutableMapOf()
+        val queue = ArrayDeque(upperBoundConcreteClassToTypes.keys)
+        while (queue.isNotEmpty()) {
+            for (superConcreteClass in queue.removeFirst().getSuperConcreteClasses(this)) {
+                subClassCounts.computeIfAbsent(superConcreteClass) {
+                    queue.addLast(it)
+                    AtomicInteger()
+                }.incrementAndGet()
+            }
+        }
```

Comment:
There can be more than 2 upper bounds, so do you mean to do n^2 number of subtyping checks?

I also added an optimization to avoid travsering class hierarchies if the given inputs are already deemed compatible before we consider subclass-relation.

Reply:
If we don't travserse the hierarchy, then we will need n^2 subtype checks right? There can be more than 2 types in question due to checking type parameters, so I didn't go down this route. BTW, I also added an optimization to avoid traversing the hierarchy if possible.

Reply:
I see... but I still feel like there could be easy yet meaningful optimizations here. E.g.,  if `superConcreteClass` we're about to visit is `in upperBoundClasses`, can we skip queueing that and potentially not waking through its hierarchy again? Of course we need to count that, though.

Reply:
Ah I see! I should have initialized `subClassCounts` with stuff from `subClassCounts` so we won't re-add things into the queue.